docs: JSON:API refactor (#472)
flarum/docs: docs/extend/api.md

use Flarum\Api\Schema;

/** @extends AbstractDatabaseResource<Label> */
class LabelResource extends AbstractDatabaseResource
{
    public function type(): string
    {
        return 'labels';
    }

    public function model(): string
    {
        return Label::class;
    }

    public function scope(Builder $query, Context $context): void
    {
        $query->whereVisibleTo($context->getActor());
    }

    public function endpoints(): array
    {
        return [
            Endpoint\Show::make(),
            Endpoint\Create::make()
                ->authenticated()
                ->can('createLabel'),
            Endpoint\Update::make()
                ->authenticated()
                ->can('edit'),
            Endpoint\Delete::make()
                ->authenticated()
                ->can('delete'),
            Endpoint\Index::make()
                ->defaultInclude(['parent']),
        ];
    }

    public function fields(): array
    {
        return [
            Schema\Str::make('name')
                ->requiredOnCreate()
                ->writable(),
            Schema\Str::make('description')
                ->writable()

>                 ->maxLength(700)

Author: good point

                ->writable()
                ->unique('labels', 'slug', true)
                ->regex('/^[^\/\\ ]*$/i'),
            Schema\Str::make('color')
                ->writable()
                ->nullable()
                ->rule('hex_color'),
            Schema\Str::make('icon')
                ->writable()
                ->nullable(),
            Schema\Boolean::make('isActive')
                ->writable(),
            Schema\DateTime::make('createdAt'),
            Schema\Boolean::make('canAddToDiscussion')
                ->get(fn (Tag $tag, FlarumContext $context) => $context->getActor()->can('addToDiscussion', $tag)),

            Schema\Relationship\ToOne::make('user')
                ->type('users')
                ->includable(),
            Schema\Relationship\ToOne::make('parent')
                ->type('labels')
                ->includable(),
            Schema\Relationship\ToMany::make('children')
                ->type('labels')
                ->includable(),
        ];
    }
    
    public function sorts(): array
    {
        return [
            SortColumn::make('createdAt'),
        ];
    }
}
```

### Resource Definition

> The API resource class must extend the `Flarum\Api\Resource\AbstractDatabaseResource` class when interacting with Eloquent models, and `Flarum\Api\Resource\AbstractResource` when not. The `type` method should return a unique string that identifies the resource type. In the case of a database resource, a model method must return the class name of the model.

Author: Not a very common way of referring to the class name 😬, how about

```suggestion
The API resource class must extend the `Flarum\Api\Resource\AbstractDatabaseResource` class when interacting with Eloquent models, and `Flarum\Api\Resource\AbstractResource` when not. The `type` method should return a unique string that identifies the resource type. In the case of a database resource, the `model` method must return the class name of the model (`::class` property).
```

    }

    public function model(): string
    {
        return Label::class;
    }
}
```

### Scoping Database Resources

The `scope` method is used to apply a query scope to the model. This is useful for applying visibility scoping and ensures no data is returned that the actor should not have access to, including when the resource is a serialized relationship of another resource.

```php
use Tobyz\JsonApiServer\Context;
use Illuminate\Database\Eloquent\Builder;

public function scope(Builder $query, Context $context): void
{
    $query->whereVisibleTo($context->getActor());
}
```

### Listing Resources

The `Index` endpoint lists the model instances. 

```php
public function endpoints(): array
{
    return [
        Endpoint\Index::make(),
    ];
}
```

:::info

Find out more about the listing endpoint in the underlying package's documentation: https://tobyzerner.github.io/json-api-server/list.html

:::

#### Pagination

You can paginate the resources being **listed** to by specifying the `limit` and `maxLimit` through the `paginate` method:

```php
use Flarum\Api\Endpoint;

public function endpoints(): array
{
    return [
        Endpoint\Index::make()

>             ->paginate(20, 50), // these are the default values, so you may omit them.

Author: just `->paginate()`, let me specifically say omit the arguments
```suggestion
            ->paginate(20, 50), // these are the default values, so you may omit these arguments.
```

    ];
}
```

#### Sorting

You can specify sort columns through the `sorts` method. For example the following will permit two sorting options: `createdAt` and `-createdAt`:

```php
use Flarum\Api\Sort\SortColumn;

public function sorts(): array
{
    return [
        SortColumn::make('createdAt'),
    ];
}
```

You can specify the default sort through the `defaultSort` method on the `Index` endpoint:

```php
use Flarum\Api\Endpoint;

public function endpoints(): array
{
    return [
        Endpoint\Index::make()
            ->defaultSort('-createdAt'),
            ->paginate(),
    ];
}
```

#### Searching and Filtering

> Read our [searching and filtering](search.md) guide for more information!

Reviewer: Unrelated to this PR, but I think docs.flarum.org might not be up to date

Author: no, because that's 1.x, and this is for the 2.x branch which is still not published until 2.0 is done ;) https://github.com/flarum/docs/pull/466

```php
use Flarum\Api\Endpoint;

public function endpoints(): array
{
    return [
        Endpoint\Show::make()
            ->authenticated()
            ->can('view'), // equivalent to $actor->can('view', $label)
        Endpoint\Create::make()
            ->authenticated()
            ->can('createLabel'), // equivalent to $actor->can('createLabel'),
        Endpoint\Update::make()
            ->admin(), // equivalent to $actor->isAdmin()
    ];
}
```

### Including relationship by default

We do not recommend including relationships by default. If possible, it is better to extend the specific request payloads to be made on the frontend side and add the includes there as that keeps the API responses optimized. However, if you *really* need to include a relationship by default, you can do so through the `defaultInclude` method:

```php
use Flarum\Api\Endpoint;

public function endpoints(): array
{
    return [
        Endpoint\Index::make()
            ->defaultInclude(['parent']),
    ];
}
```

### Lifecycle Hooks

> Some methods on the endpoint allow you to hook certain logic into the lifecycle of the endpoint. These are `before`, `after`, and `beforeSerialization` which is often not very different from `after` but is called before it when available on the endpoint.

Reviewer: Might be worth talking about what you might want to do here, and why you wouldn't do it in the database resource hooks.

```php
/** @var \Flarum\Http\UrlGenerator $url */
$url->to('api')->route('labels.index');
$url->to('api')->route('labels.activate', ['id' => $label->id]);
```

## Fields (Attributes and Relationships)

The `fields` method on the API resource is used to define the fields (attributes and relationships) of the model. You can use the `Schema` namespace to define the various field types.

### Attributes

Before you define an attribute, decide which type of attribute it is. The `Schema` namespace provides a range of attribute types, such as `Str`, `Integer`, `Boolean`, `DateTime` and `Arr` for arrays.

```php
use Flarum\Api\Schema;

public function fields(): array
{
    return [
        Schema\Str::make('name')
            ->requiredOnCreate()
            ->writable(),
        Schema\Integer::make('discussionCount'),
        Schema\Arr::make('customData'),
        Schema\Boolean::make('isActive')
            ->writable(),
        Schema\DateTime::make('createdAt'),
    ];
}
```

### Visibility

> You can use the `visible` method to determine whether an attribute should be visible in the API response.

Reviewer: ```suggestion
You can use the `visible` method to conditionally include an attribute in the API response.
```

```php
use Flarum\Api\Context;
use Flarum\Api\Schema;

> public function fields(): array

Reviewer: maybe worth noting somewhere towards the top that all these examples are of methods in the resource class.

{
    return [
        Schema\Str::make('name')
            ->visible(fn (Label $label, Context $context) => $context->getActor()->can('edit', $label)),
    ];
}
```

### Writability

> By default, a field is not writable unless you specify it to be so. You can use the `writable` method to make a field writable.

Reviewer: Can you specify custom logic for how a field will be written to the database?

EDIT: literally 2 sections below. Might be worth saying that it will write directly to the model's field by default, but will use a custom setter if provided?

    ];
}
```

### Requirability

By default, a field is not required unless you specify it to be so. You can use the methods `required`, `requiredOnCreate`, `requiredWith`, `requiredWithout`, `requiredOnCreateWith`, `requiredOnUpdateWith`, `requiredOnCreateWithout`, `requiredOnUpdateWithout`.

:::caution

You will normally only want to only require fields on creation so that fields can be updated in isolation of other attributes. So we recommend using `requiredOnCreate` methods by default unless the need for otherwise arises.

:::

```php
use Flarum\Api\Schema;

public function fields(): array
{
    return [
        Schema\Str::make('name')
            ->requiredOnCreate(),
    ];
}
```

### Getter & Setter

You can use the `get` and `set` methods to define a getter and setter for an attribute.

```php
use Flarum\Api\Context;
use Flarum\Api\Schema;

public function fields(): array
{
    return [
        Schema\Str::make('name')
            ->get(fn (Label $label) => strtoupper($label->name))
            ->set(function (Label $label, string $value, Context $context) {
                $label->name = strtolower($value);
            }),
    ];
}
```

### Validation

> You can use the `rule` method to add a [Laravel validation rule](https://laravel.com/docs/10.x/validation#available-validation-rules) to an attribute. Each type of attribute will have some custom validation rules available as methods.

Reviewer: ```suggestion
You can use the `rule` method to add a [Laravel validation rule](https://laravel.com/docs/10.x/validation#available-validation-rules) to an attribute. We've provided helper methods on some attributes for common validation rules.
```

            ->required() // required on both create and update.
            ->unique('labels', 'slug', true) // unique in the labels table, ignoring the current model.
            ->regex('/^[^\/\\ ]*$/i'), // must match the regex.
        Schema\Str::make('color')
            ->rule('hex_color'),
        Schema\Number::make('price')
            ->min(1)
            ->max(100),
        Schema\DateTime::make('createdAt')
            ->before('2022-01-01')
            ->after('2021-01-01'),
    ];
}
```

### Property

By default you should use camelCase for your attribute names and they will be automatically mapped to their snake_case equivalent when interacting with the model. But if you need to specify which property on the model the attribute should map to, you can use the `property` method.

```php
use Flarum\Api\Schema;

public function fields(): array
{
    return [
        Schema\Str::make('name')
            ->property('name_column'),
    ];
}
```

### Relationship aggregates

The `Number` & `Integer` attributes are able to get efficiently get relationship aggregates such as counts, sums, max, and min.

```php
use Flarum\Api\Schema\Number;
use Flarum\Api\Schema\Integer;

> Integer::make('commentCount')

Reviewer: Where would this code go?

    ->countRelation('comments'),

Number::make('avgRevenue')
    ->avgReation('reports', 'revenue'),

Number::make('revenueSum')
    ->sumRelation('reports', 'revenue'),

Number::make('minNumber')
    ->minRelation('posts', 'number'),

Number::make('maxNumber')
    ->maxRelation('posts', 'number'),
```

## Relationships

A relationship is a field, so all that was mentioned about fields above applies to relationships as well.

There are two types of relationships: `ToOne` and `ToMany`. You can use the `Schema\Relationship\ToOne` and `Schema\Relationship\ToMany` classes to define them.

```php
use Flarum\Api\Schema;

public function fields(): array
{
    return [
        Schema\Relationship\ToOne::make('user')
            ->type('users'),
        Schema\Relationship\ToMany::make('children')
            ->type('labels'),
    ];
}
```

### Inclusion & Linkage

> You can mark a relationship as includable through the `includable` method. This means that the relationship can be included in the API response. You can also use the `withLinkage` and `withoutLinkage` methods to determine whether the relationship ID(s) should be included in the API response (`ToMany` relationships are not linked by default contrary to `ToOne` relationships).

Reviewer: When would you have non-includable relationships?

Author: When you only need a relationship to be writable perhaps. Though not a common use case, it remains valid.

Reviewer: Might be worth adding a short tip about this

:::danger

> Adding linkage for `ToMany` relationships can lead to performance issues as it will include the IDs of all the related models in the API response. This is why it is not linked by default.

Reviewer: I think I don't fully understand linkage from this description. Is it a concept from Toby's server, or smething we added?

Author: Yes, it's not straightforward to know what it's referring to, perhaps an example would be good to have here.

Linkage is the inclusion of the relationship IDs, like:

```json
"attributes": { ... },
"relationships": { "firstPost": { "type": "posts", "id": 4 } }
```

Author: added tip explaining linkage in this section

:::

```php
use Flarum\Api\Schema;

public function fields(): array
{
    return [
        Schema\Relationship\ToOne::make('user')
            ->type('users')
            ->includable()
            ->withoutLinkage(),
    ];
}
```

### Polymorphic Relationships

> You use the `collection` method to define the resource types that a polymorphic relationship can point to.

Reviewer: Are there docs by us or on the Eloquent site we can link to re: polymorphic relationships?

Author: good point
